# Assignment 1 - Finding the Best Delivery Promise for Rosa's Pizza

#### Installing numpy and starter 

In [1]:
%pip install git+https://github.com/zhouy185/rosa-starter.git

  Cloning https://github.com/zhouy185/rosa-starter.git to /private/var/folders/dj/xjmzdbwd7jj8bk07b5lhgr680000gn/T/pip-req-build-fr9begri
  Running command git clone --filter=blob:none --quiet https://github.com/zhouy185/rosa-starter.git /private/var/folders/dj/xjmzdbwd7jj8bk07b5lhgr680000gn/T/pip-req-build-fr9begri
  Resolved https://github.com/zhouy185/rosa-starter.git to commit 09f3f0b98fd2fee1ba08322bf8db12c4940590bd
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Note: you may need to restart the kernel to use updated packages.


In [2]:
import numpy as np
from starter import ZONES, TIME_BLOCKS, COSTS, PROMISE, delivery_times

In [3]:
print(ZONES)

['Central', 'North', 'Far West']


In [4]:
print(PROMISE)

45


In [5]:
print(TIME_BLOCKS)

['Lunch', 'Weekday eve', 'Fri/Sat eve', 'Other']


In [6]:
times = delivery_times("Far West", "Fri/Sat eve", 45)
print(times)


[ 46.2  43.4  28.5  30.2  31.5  34.3  31.   30.6  71.3  56.1  37.5  72.8
  51.7  39.   35.8  32.8  39.3  34.6  36.4  41.9  44.   43.5  36.   31.4
  33.2  54.8  38.   48.1  56.6  25.9  32.9  41.9  29.8  31.8  32.2  45.4
  30.4  40.9  42.3  37.2  40.1  38.4  38.5  36.6  23.7  34.4  34.3  42.2
  54.2  36.3  38.4  38.   77.7  36.6  43.6  44.9  38.9  51.8  28.7  32.4
  44.8  34.2  40.9  28.5  26.5  40.3  49.7  48.7  34.   19.4  28.5  62.3
  26.   31.3  29.7  50.9  41.   49.8  62.5  49.4  36.6  35.6 101.6  35.
  51.1  45.9  49.   26.5  33.8  23.2  45.2  60.5  36.9  43.8  30.   37.8
  52.4  44.4  46.3  51.7  29.4  47.7  57.   35.8  40.1  57.7  40.3  32.6
  85.4  42.1  32.   26.5  30.7  45.9  45.1  62.5  31.4  41.7  35.2  37.5
  36.2  46.8  51.2  61.8  73.4  46.1  40.5  34.1  31.1  55.1  71.1  41.9
  42.7  44.6  44.3  48.9  47.6  29.7  36.2  29.8  33.9  65.   57.1  55.7
  50.5  48.5  51.   35.4  66.5  29.8  32.8  37.   33.4  63.2  60.7  33.
  38.5  40.9  34.6 103.5  35.3  26.4  24.5  59.2  28.

## Part 1

#### Part 1a

In [7]:
def percentage_lateorders(z_list, t_list, promise):

    if z_list == 'all':
        zones = ZONES
    else:
        zones = [z_list]

    if t_list == 'all':
        blocks = TIME_BLOCKS
    
    else:
        blocks = [t_list]

    times = []

    for z in zones:
        for b in blocks:
            new_times = delivery_times(z, b, promise,seed=1)
            for t in new_times:
                times.append(t)

    late = 0
    for t in times:
        if t > promise:
            late += 1

    total = len(times)
    percentage = late / total * 100

    return percentage

In [8]:
percentage_lateorders('Far West','Fri/Sat eve',45)

36.666666666666664

In [9]:
percentage_lateorders('all','Lunch',45)

3.3333333333333335

In [10]:
percentage_lateorders('Central','all',45)

4.22680412371134

#### Part 1b

In [11]:
results = [] # Creating a list for the results so I can rank them

for Z in ZONES:
    for T in TIME_BLOCKS:
        rate = percentage_lateorders(Z,T,PROMISE)
        results.append((Z,T,rate)) # This adds each loop to results

ranked = sorted(results, key=lambda x: x[2], reverse=True) # This code is to sort the results based on the rate from high to low

for rank, (zone, block, rate) in enumerate(ranked, start=1):
    print(f"{rank}. {zone} - {block}: {rate:.0f}%")

1. Far West - Fri/Sat eve: 37%
2. North - Fri/Sat eve: 17%
3. Far West - Weekday eve: 11%
4. Central - Fri/Sat eve: 7%
5. North - Weekday eve: 6%
6. Far West - Other: 5%
7. Central - Weekday eve: 4%
8. North - Lunch: 4%
9. Central - Lunch: 3%
10. Central - Other: 3%
11. North - Other: 2%
12. Far West - Lunch: 2%


#### Part 1c

In [12]:
def avg_deliverytime(z_list, t_list, promise):

    if z_list == 'all':
        zones = ZONES
    else:
        zones = [z_list]

    if t_list == 'all':
        blocks = TIME_BLOCKS
    
    else:
        blocks = [t_list]

    times = []

    for z in zones:
        for b in blocks:
            new_times = delivery_times(z, b, promise, seed=1)
            for t in new_times:
                times.append(t)

    return round(float(np.mean(times)))

In [13]:
avg_deliverytime('North','Weekday eve',45)

31

In [14]:
avg_deliverytime('all','Other',45)

28

In [15]:
avg_deliverytime('Far West','all',45)

35

In [16]:
avg_deliverytime('all','all',45)

31

#### Part 1d

In [17]:
result_2 = [] 

for place in ZONES:
    for time in TIME_BLOCKS:
        avg = avg_deliverytime(place,time,PROMISE)
        result_2.append((place,time,avg)) 
        
ranking = sorted(result_2, key=lambda x: x[2], reverse=True) # This code is to sort the results based on the rate from high to low

for ranks, (p,t,avg) in enumerate(ranking, start=1):
    print(f"{ranks}. {p} - {t}: {avg}")

1. Far West - Fri/Sat eve: 43
2. North - Fri/Sat eve: 37
3. Far West - Weekday eve: 35
4. Central - Fri/Sat eve: 32
5. North - Weekday eve: 31
6. Far West - Lunch: 31
7. Far West - Other: 30
8. Central - Weekday eve: 29
9. North - Lunch: 29
10. North - Other: 28
11. Central - Lunch: 27
12. Central - Other: 26


#### Part 1e

 ##### The output to part 1b is more relevant to Rosa's decision because it shows us the percentages of orders for each zone and time block that are outside of the delivery promise (in the example 45 mins was used). The output for part 1d shows us the average delivery time and this mean value does not necessarily show us outliers or orders that are outside of the promise time. For example the average delivery for zone: Far West, Fri/Sat eve is 44 mins (using output from part 1e). This average sits comfortably inside the promise time of 45 mins but the percentages of late orders (using output from part 1b) is 40%. So even though the average is 44 mins, there were still 40% of orders outside of the promised time. 

## Part 2

#### Part 2a

In [18]:
print(COSTS)

{'refund': 10.0, 'churn_orders': 1.8, 'margin': 9.0}


##### The code to calculate the total cost per late order is shown below

In [19]:
costs = COSTS
def total_cost(costs):
    totalcost = costs['refund'] + (costs['churn_orders'] * costs['margin'])

    return(totalcost)

In [20]:
total_cost(costs)

26.2

#### Part 2b

In [21]:
def net_profit(z, t_b, p, costs):
    order_times = delivery_times(z, t_b, p, seed=1)
    late_orders = np.sum(order_times > p)           
    sales = costs['margin'] * len(order_times)     
    late_costs = late_orders * total_cost(costs)
    return sales - late_costs

##### The codes below shows the statistics for a zone: Far West and time block: Fri/Sat eve. This will help to see where most of the deliveries fall and to determine a best promise time

In [22]:
t = delivery_times('Far West', 'Fri/Sat eve', PROMISE, seed=1) 

print(f"Minimum:          {np.min(t):.1f} min")
print(f"10th percentile:  {np.percentile(t, 10):.1f} min")
print(f"Average:          {np.mean(t):.1f} min")
print(f"90th percentile:  {np.percentile(t, 90):.1f} min")
print(f"Maximum:          {np.max(t):.1f} min")

Minimum:          18.6 min
10th percentile:  29.3 min
Average:          42.9 min
90th percentile:  55.4 min
Maximum:          99.4 min


In [23]:
promises = list(range(5, 55, 5)) # This increases the promise time in steps of 5 and stops at 55
table = []

print(f"{'Promise':>8} {'Orders':>8} {'Late %':>8} {'Net profit':>12}")
for p in promises:
    t = delivery_times('Far West', 'Fri/Sat eve', p, seed=1)
    profit = net_profit('Far West', 'Fri/Sat eve', p, COSTS)
    table.append((p, profit))
    print(f"{p:>8} {len(t):>8} {np.mean(t > p) * 100:>7.0f}% {profit:>12,.2f}")

 Promise   Orders   Late %   Net profit
       5      242     100%    -4,162.40
      10      242     100%    -4,162.40
      15      241     100%    -4,145.20
      20      241      99%    -4,092.80
      25      239      97%    -3,953.60
      30      237      92%    -3,578.60
      35      232      77%    -2,601.80
      40      224      59%    -1,468.60
      45      210      37%      -127.40
      50      188      18%       827.40


##### Using 'Far West' and 'Fri/Sat eve', the statistics above shows that the fastest delivery takes 18.6 minutes and only 10% of orders arrive within 29.3 minutes or less. Promise time shorter than about 25 minutes are therefore unrealistic because over 97% of orders would be late, and Rosa would lose more than $3,900. This is why the lower end of the range does not need to go below 25 minutes, and the useful promises start closer to 40 minutes.

In [24]:
promises = list(range(5, 85, 5))

def best_promise(z, t_b, promise_list, costs):
       best_p = None
       best_profit = float('-inf')
       for p in promise_list:
           profit = net_profit(z, t_b, p, costs)
           if profit > best_profit:
               best_profit = profit
               best_p = p
       return best_p, best_profit

##### The range chosen is from 5 to 80 minutes because. This is because the lower end is kept low so the function also works for faster zones and time blocks. The upper end of the range (from 60 - 80) is well beyond the 90th percentile of delivery times for the slowest pair (55.4 minutes for Far West, Fri/Sat eve). The edge check confirms that no pair's best promise falls at either end of the range.

In [25]:
p, profit = best_promise('Far West','Fri/Sat eve', promises, COSTS)
print(f"Recommended promise: {p} minutes")
print(f"Net profit: ${profit:,.2f}")

Recommended promise: 55 minutes
Net profit: $1,212.40


##### For zone Far West, during Fri/Sat eve time block, the best promise time was calculated to be 55 minutes. 

In [26]:
for z in ZONES:
    for b in TIME_BLOCKS:
        p, profit = best_promise(z, b, promises, COSTS)
        if p == promises[0] or p == promises[-1]:
            note = "AT EDGE: widen the range"
        else:
            note = "inside range: OK"
        print(f"{z} - {b}: best {p} min, net profit ${profit:,.2f} ({note})")

Central - Lunch: best 45 min, net profit $2,104.20 (inside range: OK)
Central - Weekday eve: best 45 min, net profit $2,603.20 (inside range: OK)
Central - Fri/Sat eve: best 50 min, net profit $1,399.00 (inside range: OK)
Central - Other: best 45 min, net profit $1,579.00 (inside range: OK)
North - Lunch: best 45 min, net profit $1,526.60 (inside range: OK)
North - Weekday eve: best 45 min, net profit $1,857.00 (inside range: OK)
North - Fri/Sat eve: best 50 min, net profit $996.40 (inside range: OK)
North - Other: best 45 min, net profit $1,335.20 (inside range: OK)
Far West - Lunch: best 45 min, net profit $1,001.40 (inside range: OK)
Far West - Weekday eve: best 50 min, net profit $1,210.80 (inside range: OK)
Far West - Fri/Sat eve: best 55 min, net profit $1,212.40 (inside range: OK)
Far West - Other: best 45 min, net profit $1,492.20 (inside range: OK)


##### All 12 zone and time block pairs have their best promise inside the 5 – 80 minute range, so the range is wide enough to capture the most profitable promise for every pair.